cli/util/eta.go

```diff
@@ -12,46 +12,58 @@ type ETA struct {
 	// max number of items to keep in memory
 	maxItems int
 	// a queue of most recently updated items
-	items []time.Time
+	items []item
 	// we store the last calculated ETA which we reuse if there was not change in remaining items
 	lastETA string
-	// we store the last remaining value we processed so we can ignore updates with the same value
-	lastRemaining int64
+}
+
+type item struct {
+	timestamp time.Time
+	remaining int64
 }
 
 // NewETA creates a new ETA calculator of the given size
 func NewETA(maxItems int) *ETA {
 	return &ETA{
 		maxItems: maxItems,
-		items:    make([]time.Time, 0),
+		items:    make([]item, 0),
 	}
 }
 
 // Update updates the ETA calculator with the remaining number of items and returns the ETA
 func (e *ETA) Update(remaining int64) string {
-	ts := time.Now()
+	item := item{
+		timestamp: time.Now(),
+		remaining: remaining,
+	}
 
 	if len(e.items) == 0 {
-		e.items = append(e.items, ts)
-		e.lastRemaining = remaining
+		e.items = append(e.items, item)
 		return ""
 	}
 
-	// we ignore updates with the same remaining value and just return the previous ETA
-	if e.lastRemaining == remaining {
+	if e.items[len(e.items)-1].remaining == remaining {
+		// we ignore updates with the same remaining value and just return the previous ETA
 		return e.lastETA
+	} else if e.items[len(e.items)-1].remaining < remaining {
+		// in case remaining went up from previous updates then we don't know how many items
+		// were actually updated, lets assume 1 and update the remaining value of all items
```

Comment:
This works but it does introduce an annoying "blip". Can we extrapolate from the current rate instead? That'll keep the rate the same.